src/nnet3/nnet-simple-component.cc

```diff
@@ -1011,6 +1012,185 @@ void PerElementScaleComponent::UnVectorize(
   scales_.CopyFromVec(params);
 }
 
+void PerElementOffsetComponent::Scale(BaseFloat scale) {
+  offsets_.Scale(scale);
+}
+
+void PerElementOffsetComponent::Resize(int32 dim) {
+  KALDI_ASSERT(dim > 0);
+  offsets_.Resize(dim);
+}
+
+void PerElementOffsetComponent::Add(BaseFloat alpha,
+                                   const Component &other_in) {
+  const PerElementOffsetComponent *other =
+      dynamic_cast<const PerElementOffsetComponent*>(&other_in);
+  KALDI_ASSERT(other != NULL);
+  offsets_.AddVec(alpha, other->offsets_);
+}
+
+PerElementOffsetComponent::PerElementOffsetComponent(
+    const PerElementOffsetComponent &component):
+    UpdatableComponent(component),
+    offsets_(component.offsets_) { }
+
+void PerElementOffsetComponent::SetZero(bool treat_as_gradient) {
+  if (treat_as_gradient) {
+    SetLearningRate(1.0);
+    is_gradient_ = true;
+  }
+  offsets_.SetZero();
+}
+
+void PerElementOffsetComponent::PerturbParams(BaseFloat stddev) {
+  CuVector<BaseFloat> temp_offsets(offsets_);
+  temp_offsets.SetRandn();
+  offsets_.AddVec(stddev, temp_offsets);
+}
+
+std::string PerElementOffsetComponent::Info() const {
+  std::stringstream stream;
+  BaseFloat offsets_mean = offsets_.Sum() / offsets_.Dim();
+  BaseFloat offsets_stddev = std::sqrt(VecVec(offsets_, offsets_) /
+                              offsets_.Dim()) - (offsets_mean * offsets_mean);
+  stream << Type() << ", input-dim=" << InputDim()
+         << ", output-dim=" << OutputDim()
+		 << ", offsets-mean=" << offsets_mean
+         << ", offsets-stddev=" << offsets_stddev
+         << ", learning-rate=" << LearningRate();
+  return stream.str();
```

Comment:
There is nothing to stop that, but it's probably unlikely in practice.  You
could have min, max and min-abs.


On Tue, Nov 3, 2015 at 4:09 PM, hhadian <notifications@github.com> wrote:

> In src/nnet3/nnet-simple-component.cc
> <https://github.com/kaldi-asr/kaldi/pull/307#discussion_r43807143>:
>
> > +  CuVector<BaseFloat> temp_offsets(offsets_);
> > +  temp_offsets.SetRandn();
> > +  offsets_.AddVec(stddev, temp_offsets);
> > +}
> > +
> > +std::string PerElementOffsetComponent::Info() const {
> > +  std::stringstream stream;
> > +  BaseFloat offsets_mean = offsets_.Sum() / offsets_.Dim();
> > +  BaseFloat offsets_stddev = std::sqrt(VecVec(offsets_, offsets_) /
> > +                              offsets_.Dim()) - (offsets_mean * offsets_mean);
> > +  stream << Type() << ", input-dim=" << InputDim()
> > +         << ", output-dim=" << OutputDim()
> > +		 << ", offsets-mean=" << offsets_mean
> > +         << ", offsets-stddev=" << offsets_stddev
> > +         << ", learning-rate=" << LearningRate();
> > +  return stream.str();
>
> Can scales in scale-component be negative? If yes then I think min and max
> would be more useful if they are in terms of magnitude; otherwise we might
> miss near-zero scales.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/307/files#r43807143>.
>